torch_xla_py/data_parallel.py

```diff
@@ -172,8 +173,9 @@ def __init__(self, network, device_ids=None, batchdim=0, drop_last=False):
         xm.Replication(self._device_ids, replication_devices)
         if replication_devices else None)
     self._models = []
+    netw = network()
```

Comment:
Good catch Taylan!
But let's do a change, since it came up with SF as well.
```
module = network() if iscallable(network) else network
for device in device_ids:
  devive_module = deepcopy(module).to(device=torch.device(device))
  self._models.append(device_module)
```

Reply:
Actually, you do not need deepcopy. As long as the model is generated once outside the loop, we are fine:
```
module = network() if iscallable(network) else network
for device in device_ids:
  devive_module = module.to(device=torch.device(device))
  self._models.append(device_module)
```

The `to()` call will simply copy the same parameters to the different devices.

Reply:
Actually2 😁 
The network (torch.nn.Module) as callable, so:
```
module = network if isinstance(network, torch.nn.Module) else network()
for device in device_ids:
  devive_module = module.to(device=torch.device(device))
  self._models.append(device_module)
```

Reply:
Interesting. We were already calling network() before, so I guess this is a separate fix, two birds with 1 stone.

Reply:
@dlibenzi I tested this version, and it actually does not work w/o deepcopy.

```
Exception in model function for device=xla:8: Function AddmmBackward returned an invalid gradient at index 0 - expected device xla:1 but got xla:8
Exception in model function for device=xla:7: Function AddmmBackward returned an invalid gradient at index 0 - expected device xla:1 but got xla:7
Exception in model function for device=xla:1: Function AddmmBackward returned an invalid gradient at index 2 - expected device xla:8 but got xla:1
Exception in model function for device=xla:5: Function AddmmBackward returned an invalid gradient at index 0 - expected device xla:1 but got xla:5
Exception in model function for device=xla:6: Function AddmmBackward returned an invalid gradient at index 0 - expected device xla:1 but got xla:6
Exception in model function for device=xla:2: Function AddmmBackward returned an invalid gradient at index 0 - expected device xla:1 but got xla:2
Exception in model function for device=xla:4: Function AddmmBackward returned an invalid gradient at index 0 - expected device xla:1 but got xla:4
Exception in model function for device=xla:3: Function AddmmBackward returned an invalid gradient at index 0 - expected device xla:1 but got xla:3
Traceback (most recent call last):
  File "/pytorch/xla/torch_xla_py/data_parallel.py", line 204, in _module_runner
    result.result = loop_fn(module, loader, torch.device(device), context)
Traceback (most recent call last):
  File "/pytorch/xla/torch_xla_py/data_parallel.py", line 204, in _module_runner
    result.result = loop_fn(module, loader, torch.device(device), context)
  File "/pytorch/xla/test/test_train_mnist.py", line 103, in train_loop_fn
    loss.backward()
  File "/root/anaconda3/envs/pytorch/lib/python3.6/site-packages/torch/tensor.py", line 120, in backward
    torch.autograd.backward(self, gradient, retain_graph, create_graph)
  File "/root/anaconda3/envs/pytorch/lib/python3.6/site-packages/torch/autograd/__init__.py", line 93, in backward
    allow_unreachable=True)  # allow_unreachable flag
Traceback (most recent call last):
  File "/pytorch/xla/torch_xla_py/data_parallel.py", line 204, in _module_runner
    result.result = loop_fn(module, loader, torch.device(device), context)
Traceback (most recent call last):
  File "/pytorch/xla/test/test_train_mnist.py", line 103, in train_loop_fn
    loss.backward()
  File "/root/anaconda3/envs/pytorch/lib/python3.6/site-packages/torch/tensor.py", line 120, in backward
    torch.autograd.backward(self, gradient, retain_graph, create_graph)
Traceback (most recent call last):
  File "/root/anaconda3/envs/pytorch/lib/python3.6/site-packages/torch/autograd/__init__.py", line 93, in backward
    allow_unreachable=True)  # allow_unreachable flag
RuntimeError: Function AddmmBackward returned an invalid gradient at index 0 - expected device xla:1 but got xla:3
Traceback (most recent call last):
  File "/pytorch/xla/torch_xla_py/data_parallel.py", line 204, in _module_runner
    result.result = loop_fn(module, loader, torch.device(device), context)
Traceback (most recent call last):
Traceback (most recent call last):
  File "/pytorch/xla/torch_xla_py/data_parallel.py", line 204, in _module_runner
    result.result = loop_fn(module, loader, torch.device(device), context)
```

Reply:
Hmm, module.to() should create a new model with all parameters moved.
Can you debug a little?